### PR TITLE
feat: improve guild creation, add new guild builder

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Adds support for the previously missing parameters on the [create guild endpoint](https://discord.com/developers/docs/resources/guild#create-guild) (literally all of them except `name` + `icon`), adding a builder class to abstract away most of the placeholder ID handling.

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by onerandomusername** [comment]
Sorry, misclicked

File: docs/api.rst [hunk 1]
```diff
@@ -4470,6 +4470,15 @@ Guild
...
         :type: :class:`User`
 
+GuildBuilder
+~~~~~~~~~~~~~
+
+.. attributetable:: GuildBuilder
+
+.. autoclass:: GuildBuilder()
+    :members:
+    :exclude-members: add_category_channel
```
Comment (EmreTech) [inline docs/api.rst:4480]: Why is `add_category_channel` excluded?

File: disnake/guild.py [hunk 2]
```diff
@@ -4191,3 +4204,464 @@ async def timeout(
...
+        if user_limit is not MISSING:
+            data["user_limit"] = user_limit
+
+        if rtc_region is not MISSING:
+            data["rtc_region"] = None if rtc_region is None else str(rtc_region)
+
+        if video_quality_mode is not MISSING:
+            data["video_quality_mode"] = try_enum_to_int(video_quality_mode)
+
+        return _id
```
Comment (EmreTech) [inline disnake/guild.py:4667]: Any plans for an `add_stage_channel` function?

**Review by EmreTech — COMMENTED** [review]
Looks great otherwise!

File: disnake/guild.py (hunk 2 above)
Comment (onerandomusername) [reply disnake/guild.py:4667]: I don't think they can be created. Stage channels are only for community guilds and as far as I can tell a guild created cannot be set to COMMUNITY on creation.

Comment (EmreTech) [reply disnake/guild.py:4667]: Ah good point

Comment (shiftinv) [reply disnake/guild.py:4667]: Yup, tried that as well. It doesn't look like you can set features on creation, and the only allowed channel types are 0 (text), 2 (voice), and 4 (category).

File: docs/api.rst (hunk 1 above)
Comment (shiftinv) [reply docs/api.rst:4480]: It's an alias of `add_category` - since sphinx can't properly detect method aliases, it'd otherwise show the entire docstring twice :/

Comment (shiftinv) [reply docs/api.rst:4480]: Perhaps, but it'll look out of place as sphinx wouldn't put it next to `add_category` in the docs, but instead all the way at the top of the method list. I've added a mention of `add_category_channel` to the `add_category` docs now, that should do it as well.
91ce8dd8d621be24bf114f590dee96aa98291579

**Review by EmreTech — APPROVED** [review]

**Review by onerandomusername — COMMENTED** [review]
unfinished review

File: changelog/578.feature.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Add :class:`GuildBuilder` for full coverage of the guild creation endpoint.
```
[the remark this answers is not in the document]
Comment (shiftinv) [reply changelog/578.feature.rst]: https://github.com/DisnakeDev/disnake/pull/578/commits/cb8bea7b9d18bfd901c66edff35e40cde68dbdf4

File: disnake/client.py [hunk 4]
```diff
@@ -1789,6 +1790,26 @@ async def create_guild(
...
+        Parameters
+        ----------
+        name: :class:`str`
+            The name of the guild.
+
+        Returns
+        -------
+        :class:`.GuildBuilder`
+            The guild builder object for configuring and creating a new guild.
+        """
```
[the remark this answers is not in the document]
Comment (shiftinv) [reply disnake/client.py]: https://github.com/DisnakeDev/disnake/pull/578/commits/a7a61d389dcba5df1f380bc4d9d56590999d19d8

File: disnake/guild.py [hunk 5]
```diff
@@ -4553,3 +4566,468 @@ async def create_automod_rule(
...
         )
         return AutoModRule(data=data, guild=self)
+
+
+PlaceholderID = NewType("PlaceholderID", int)
+
+
+class GuildBuilder:
+    """
+    A guild builder object, created by :func:`Client.guild_builder`.
```
Comment (Victorsitou) [inline disnake/guild.py:4576]: Generally, this shouldn't have an enter, but this hasn't been consistent lately, mostly with new classes (I'll probably create a PR to change all that)

Make sure to change it in other methods below 😄

File: disnake/client.py [hunk 6]
```diff
@@ -1952,6 +1953,28 @@ async def create_guild(
...
 
+    def guild_builder(self, name: str) -> GuildBuilder:
+        """Creates a builder object that can be used to create more complex guilds.
+
+        This is a more comprehensive alternative to :func:`create_guild`.
+        See :class:`.GuildBuilder` for details and examples.
+
+        Bot accounts in 10 or more guilds are not allowed to create guilds.
+
+        .. versionadded:: 2.7
```
Comment (Victorsitou) [inline disnake/client.py:1964]: This is obvious, but well, change this to 2.8 :)

File: disnake/guild.py [hunk 7]
```diff
@@ -4553,3 +4566,468 @@ async def create_automod_rule(
...
+                permissions will *not* be synced to this new channel by default.
+
+        topic: Optional[:class:`str`]
+            The channel's topic.
+        slowmode_delay: :class:`int`
+            Specifies the slowmode rate limit for users in this channel, in seconds.
+            A value of ``0`` disables slowmode. The maximum value possible is ``21600``.
+            If not provided, slowmode is disabled.
+        nsfw: :class:`bool`
+            Whether to mark the channel as NSFW or not.
```
Comment (Victorsitou) [inline disnake/guild.py:4935]: This is optional, but in #275, I changed all "Whether ......." documentation and leave it without "or not"/"not" (there are still some documentation with that), but it'd be nice to remove "or not".

(same below)

File: disnake/guild.py [hunk 8]
```diff
@@ -4553,3 +4566,468 @@ async def create_automod_rule(
...
+            name=name,
+            overwrites=overwrites,
+            category=category,
+            topic=topic,
+            slowmode_delay=slowmode_delay,
+            nsfw=nsfw,
+        )
+
+        if default_auto_archive_duration is not MISSING:
+            data["default_auto_archive_duration"] = cast(
```
Comment (Victorsitou) [inline disnake/guild.py:4956]: Why is this here and not in `_add_channel`?

File: disnake/types/guild.py [hunk 9]
```diff
@@ -161,3 +161,25 @@ class RolePositionUpdate(TypedDict):
...
 class MFALevelUpdate(TypedDict):
     level: MFALevel
+
+
+class CreateGuildPlaceholderRole(CreateRole):
+    id: Snowflake
+
+
+class CreateGuildPlaceholderChannel(CreateGuildChannel, total=False):
+    id: Snowflake
```
Comment (Victorsitou) [inline disnake/types/guild.py:171]: Why the `total=False`?

**Review by Victorsitou — COMMENTED** [review]
I haven't tested it, but I love the idea 👍

File: disnake/guild.py (hunk 5 above)
Comment (Victorsitou) [reply disnake/guild.py:4576]: Okay never mind, just saw #722 and this can be changed in that PR

File: disnake/client.py (hunk 6 above)
Comment (shiftinv) [reply disnake/client.py]: Good catch, this PR was quite far behind, hadn't updated it since :sweat_smile:
https://github.com/DisnakeDev/disnake/pull/578/commits/75f9bf54fbad336b3fe4d0ec00f0268c544473ba

File: disnake/guild.py (hunk 7 above)
Comment (shiftinv) [reply disnake/guild.py]: https://github.com/DisnakeDev/disnake/pull/578/commits/fb57ad630e91953b2dc693cf0aa512abc725be77

File: disnake/types/guild.py (hunk 9 above)
Comment (shiftinv) [reply disnake/types/guild.py]: `id` is optional, the endpoint doesn't require placeholder IDs to be specified. I've changed it to use `NotRequired` in https://github.com/DisnakeDev/disnake/pull/578/commits/4bc79c0ef9d78b5cea1d2595e1ec92acd28a95c6 to keep it consistent with the base type though.

File: disnake/guild.py (hunk 8 above)
Comment (shiftinv) [reply disnake/guild.py:4970]: text channels are the only ones that support this field at the moment, the fields implemented in `_add_channel` are generally there to avoid duplicating code

Comment (Victorsitou) [reply disnake/guild.py:4970]: Makes sense 👍

File: disnake/guild.py [hunk 10]
```diff
@@ -4567,3 +4580,468 @@ async def create_automod_rule(
...
+        if hoist is not MISSING:
+            data["hoist"] = hoist
+
+        if mentionable is not MISSING:
+            data["mentionable"] = mentionable
+
+        self._roles.append(data)
+        return _id
+
+    def _add_channel(
```
Comment (Victorsitou) [inline disnake/guild.py:4839]: Would be nice to move this below _next_id to avoid having private methods between the public ones

**Review by Victorsitou — APPROVED** [review]

File: disnake/guild.py (hunk 10 above)
Comment (shiftinv) [reply disnake/guild.py]: sounds good.
https://github.com/DisnakeDev/disnake/pull/578/commits/e9021ef55066e609c67ea06f505f5fbe91f4f89f

File: disnake/guild.py (hunk 8 above)
Comment (Victorsitou) [reply disnake/guild.py:4970]: you can mark this as resolved

File: disnake/guild.py (hunk 5 above)
Comment (Victorsitou) [reply disnake/guild.py:4590]: you can mark this as resolved too

**Comment by shiftinv** [comment]
Tested everything again, this should be good to go :tada:
